kotlinx-coroutines-core/common/src/flow/Channels.kt

```diff
@@ -7,14 +7,81 @@
+ * the channel afterwards. If you need to iterate over the channel without consuming it,
+ * a regular `for` loop should be used instead.
+ *
+ * This function provides a more efficient shorthand for `channel.consumeEach { value -> emit(value) }`.
+ * See [consumeEach][ReceiveChannel.consumeEach].
+ */
+@ExperimentalCoroutinesApi
+public suspend fun <T> FlowCollector<T>.emitAll(channel: ReceiveChannel<T>) {
+    // Manually inlined "consumeEach" implementation that does not use iterator but works via "receiveOrClosed".
+    // It has smaller and more efficient spilled state which also allows to implement a manual kludge to
+    // fix retention of the last emitted value.
+    // See https://youtrack.jetbrains.com/issue/KT-16222
+    // See https://github.com/Kotlin/kotlinx.coroutines/issues/1333
+    var cause: Throwable? = null
+    try {
+        while (true) {
+            // :KLUDGE: This "run" call is resolved to an extension function "run" and forces the size of
+            // spilled state to increase by an additional slot, so there are 4 object local variables spilled here
+            // which makes the size of spill state equal to the 4 slots that are spilled around subsequent "emit"
+            // call, ensuring that the previously emitted value is not retained in the state while receiving
+            // the next one.
+            //     L$0 <- this
+            //     L$1 <- channel
+            //     L$2 <- cause
+            //     L$3 <- this$run (actually equal to this)
+            val result = run { channel.receiveOrClosed() }
+            if (result.isClosed) {
+                result.closeCause?.let { throw it }
+                break // returns normally when result.closeCause == null
+            }
+            // result is spilled here to the coroutine state and retained after the call, even though
+            // it is not actually needed in the next loop iteration.
+            //     L$0 <- this
+            //     L$1 <- channel
+            //     L$2 <- cause
+            //     L$3 <- result
+            emit(result.value)
+        }
+    } catch (e: Throwable) {
+        cause = e
+        throw e
+    } finally {
+        channel.cancelConsumed(cause)
+    }
+}
+
+/**
+ * Represents the given receive channel as a hot flow and [consumes][ReceiveChannel.consume] the channel
+ * on the first collection from this flow. The resulting flow can be collected just once and throws
+ * [IllegalStateException] when trying to collect it more than once.
+ *
+ * ### Cancellation semantics
+ * 1) Flow consumer is cancelled when the original channel is cancelled.
+ * 2) Flow consumer completes normally when the original channel completes (~is closed) normally.
+ * 3) If the flow consumer fails with an exception, channel is cancelled.
+ *
+ */
+@FlowPreview
+public fun <T> ReceiveChannel<T>.consumeAsFlow(): Flow<T> = object : Flow<T> {
```

Comment:
[optional] I find out that reading flow code is really unpleasant.
E.g. this method adds a frame named `kotlinx.coroutines.flow.FlowKt__ChannelsKt$consumeAsFlow$1.collect(Channels.kt:81)`. Not the best user experience :) 

While we can avoid such frames for inline methods, we can at least do a best-effort on non-inline ones

Reply:
How can it be fixed? Move the code to a named class?